packages/xforms-engine/src/client/RankNode.ts

```diff
@@ -0,0 +1,50 @@
+import type { RankType, RankControlDefinition } from '../parse/body/control/RankControlDefinition.ts';
+import type { LeafNodeDefinition } from '../parse/model/LeafNodeDefinition.ts';
+import type { BaseValueNode, BaseValueNodeState } from './BaseValueNode.ts';
+import type { RootNode } from './RootNode.ts';
+import type { TextRange } from './TextRange.ts';
+import type { GeneralParentNode } from './hierarchy.ts';
+import type { LeafNodeValidationState } from './validation.ts';
+
+export interface RankItem {
+	get label(): TextRange<'item-label'>;
+	get value(): string;
+}
+
+export type RankValueOptions = readonly RankItem[];
+
+export interface RankNodeState extends BaseValueNodeState<readonly string[]> {
+	get valueOptions(): RankValueOptions;
+
+	/**
+	 * An ordered collection of values from {@link RankItem}.
+	 * The order of the items is important and must be preserved during processing.
+	 */
+	get value(): readonly string[];
+}
+
+export interface RankDefinition extends LeafNodeDefinition<V> {
+	readonly bodyElement: RankControlDefinition;
+}
+
+export interface RankNode extends BaseValueNode<'string', readonly string[]> {
+	readonly nodeType: RankType;
+	readonly valueType: 'string';
+	readonly definition: RankDefinition<'string'>;
+	readonly root: RootNode;
+	readonly parent: GeneralParentNode;
+	readonly currentState: RankNodeState;
+	readonly validationState: LeafNodeValidationState;
+
+	/**
+	 * Convenience API to get the {@link RankItem}'s label.
+	 */
+	getValueLabel(value: string): string;
```

Comment:
```suggestion
	getValueLabel(value: string): TextRange<'item-label'> | null;
```

You wouldn't know this without background knowledge of the design intent, but `TextRange` is a stub type we'll use to provide support for ODK XForms' subset of Markdown/styling. While the support is incomplete, I think we should continue to use the same interface everywhere formatting will be applied when support is added. For now, clients reference the `asString` property to work around this incompleteness.

(I'll leave a corresponding suggestion in `RankControl.ts`.)

Reply:
Oh thanks for explaining, I didn't catch it like that before! Fixed.